components/atom/tooltip/src/index.js

```diff
@@ -0,0 +1,234 @@
+const CLASS_TARGET = `${BASE_CLASS}-target`
+
+const PLACEMENTS = {
+  TOP: 'top',
+  TOP_START: 'top-start',
+  TOP_END: 'top-end',
+  RIGHT: 'right',
+  RIGHT_START: 'right-start',
+  RIGHT_END: 'right-end',
+  BOTTOM: 'bottom',
+  BOTTOM_START: 'bottom-start',
+  BOTTOM_END: 'bottom-end',
+  LEFT: 'left',
+  LEFT_START: 'left-start',
+  LEFT_END: 'left-end'
+}
+
+class AtomTooltip extends Component {
+  state = {isOpen: false}
+  preventNonTouchEvents = false
+  hasTouchEnded = false
+  touchTimer = null
+  refTooltip = React.createRef()
+  refTarget = React.createRef()
+
+  static defaultProps = {
+    isVisible: true
+  }
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    if (!nextProps.isVisible && prevState.isOpen) {
+      return {isOpen: false}
+    }
+    return null
+  }
+
+  extendChildren() {
+    const {children} = this.props // eslint-disable-line react/prop-types
+
+    const ref = this.refTarget
+    const className = CLASS_TARGET
+    const onTouchEnd = this.toggle
+
+    const childrenOnly = React.Children.only(children)
+
+    return React.Children.map(childrenOnly, child => {
+      this.onClickTarget = child.props.onClick
+      this.title = child.props.title
+      return React.cloneElement(child, {
+        ref,
+        onClick: null,
+        className,
+        onTouchEnd
+      })
+    })
+  }
+
+  componentDidMount() {
+    const target = this.refTarget.current
+    this.props.innerRef(target)
+    document.addEventListener('click', this.handleClickOutsideElement)
+    target.oncontextmenu = this.handleContextMenu
+    target.addEventListener('mouseover', this.disableTitle)
+    target.addEventListener('mouseout', this.restoreTitle)
+  }
+
+  componentWillUnmount() {
+    const target = this.refTarget.current
+    clearTimeout(this.touchTimer)
+    document.removeEventListener('click', this.handleClickOutsideElement)
+    this.refTarget.removeEventListener('touchend', this.toggle)
+    target.removeEventListener('mouseover', this.disableTitle)
+    target.removeEventListener('mouseout', this.restoreTitle)
+  }
+
+  disableTitle(e) {
+    this.dataset.title = this.title
+    this.title = ''
+  }
+
+  restoreTitle(e) {
+    this.title = this.dataset.title
+  }
+
+  handleContextMenu = e => {
+    e.preventDefault()
+    e.stopPropagation()
+    return false
+  }
+
+  handleClickOutsideElement = e => {
+    const {isOpen} = this.state
+    if (isOpen) {
+      const tooltipDom = this.refTooltip.current
+      const isOutside = tooltipDom && !tooltipDom.contains(e.target)
+      if (isOutside) this.toggle(e)
+    }
+  }
+
+  handleTouchStart = e => {
+    this.preventNonTouchEvents = true
+    this.hasTouchEnded = false
+    clearTimeout(this.touchTimer)
+    this.touchTimer = setTimeout(() => {
+      if (!this.hasTouchEnded) {
+        this.setState({
+          isOpen: !this.state.isOpen
+        })
+      }
+      this.preventNonTouchEvents = false
+      this.hasTouchEnded = false
+    }, 1000)
```

Comment:
should this 1000 be a prop?

Reply:
Don't think so. This exact value is defined in the specification.
Anyway, it can be modified to be a prop without too much effort

Reply:
Added as prop